pkg/controller/flinkapplication/flink_state_machine.go

```diff
@@ -867,6 +916,192 @@ func (s *FlinkStateMachine) compareAndUpdateError(application *v1beta1.FlinkAppl
 
 }
 
+func getUpdatingVersion(application *v1beta1.FlinkApplication) v1beta1.FlinkApplicationVersion {
+	if getDeployedVersion(application) == v1beta1.BlueFlinkApplication {
+		return v1beta1.GreenFlinkApplication
+	}
+
+	return v1beta1.BlueFlinkApplication
+}
+
+func getDeployedVersion(application *v1beta1.FlinkApplication) v1beta1.FlinkApplicationVersion {
+	// First deploy, set the version to Blue
+	if application.Status.DeployVersion == "" {
+		application.Status.DeployVersion = v1beta1.BlueFlinkApplication
+	}
+	return application.Status.DeployVersion
+}
+
+func getCancelFlag(app *v1beta1.FlinkApplication) bool {
+	if v1beta1.IsBlueGreenDeploymentMode(app.Spec.DeploymentMode) && app.Status.Phase != v1beta1.FlinkApplicationDeleting {
+		return false
+	}
+	return true
+}
+
+// Two applications are running in this phase. This phase is only ever reached when the
+// DeploymentMode is set to BlueGreen
+func (s *FlinkStateMachine) handleDualRunning(ctx context.Context, application *v1beta1.FlinkApplication) (bool, error) {
+	if application.Spec.Teardown {
+		s.flinkController.LogEvent(ctx, application, corev1.EventTypeNormal, "TeardownInitated",
+			fmt.Sprintf("Tearing down application with hash %s and version %v", application.Status.DeployHash,
+				application.Status.DeployVersion))
+		s.updateApplicationPhase(application, v1beta1.FlinkApplicationTeardown)
+		return statusChanged, nil
+	}
+
+	// Update status of the cluster
+	hasClusterStatusChanged, clusterErr := s.flinkController.CompareAndUpdateClusterStatus(ctx, application, application.Status.DeployHash)
+	if clusterErr != nil {
+		logger.Errorf(ctx, "Updating cluster status failed with %v", clusterErr)
+		return statusUnchanged, clusterErr
+	}
+
+	// Update status of jobs on the cluster
+	hasJobStatusChanged, jobsErr := s.flinkController.CompareAndUpdateJobStatus(ctx, application, application.Status.DeployHash)
+	if jobsErr != nil {
+		logger.Errorf(ctx, "Updating jobs status failed with %v", jobsErr)
+		return statusUnchanged, jobsErr
+	}
+
+	// Update k8s object if either job or cluster status has changed
+	if hasJobStatusChanged || hasClusterStatusChanged {
+		return statusChanged, nil
+	}
+	return statusUnchanged, nil
+}
+
+func (s *FlinkStateMachine) handleTeardown(ctx context.Context, application *v1beta1.FlinkApplication) (bool, error) {
```

Comment:
There are a couple of things with the teardown phase that I'd like opinions on:
1) The Teardown spec. Right now, I've modeled it exactly like the `forceRollback` flag (which is a boolean). However, I've been thinking that this may have to actually be the version we need to teardown, simply because in the event of a bad update, the user may actually want to teardown the new(er) version. 

2) The act of teardown: Right now the cluster is torn down immediately. Should this instead try to savepoint/force cancel the job? I'm not too sure what the benefit of savepointing during a teardown would be, but force-cancelling perhaps would help in a more graceful deletion of the job/cluster.

Let me know what you think!

Reply:
1. Commented a bit about this with a slightly different concern (that people will forget to unset it, leading to unintentional tear-downs). But the one you bring up is also important: if the green deploy is bad, we need to give users a way to tear down that one.
2. I think this definitely should be a ForceCancel, so that we don't cause ZK issues.